zarrs_filesystem/src/lib.rs

```diff
@@ -270,31 +279,70 @@ impl ReadableStorageTraits for FilesystemStore {
+                    let file_size = usize::try_from(file.metadata().unwrap().size()).unwrap(); // FIXME: unwraps
+                    let ps: usize = page_size::get();
+                    let (offset, length) = match byte_range {
+                        ByteRange::FromStart(offset, option_length) => {
+                            let usize_offset = usize::try_from(offset).unwrap();
+                            match option_length {
+                                Some(length) => (usize_offset, usize::try_from(length).unwrap()),
+                                None => (usize_offset, file_size - usize_offset)
+                            }
+                        },
+                        ByteRange::Suffix(offset) => {
+                            let usize_offset = usize::try_from(offset).unwrap();
+                            (file_size - usize_offset, usize_offset)
                         }
+                    };
+                    if length > file_size {
+                        return Err(StorageError::IOError(Arc::new(std::io::Error::new(std::io::ErrorKind::UnexpectedEof, "TODO: To make test pass and match the behavior in the non-direct_io case, requesting length > file size is not permitted"))));
+                    }
+                    let fd = file.as_raw_fd();
+                    let aligned_offset = offset - (offset % ps);
+                    // If the requested length is less than the page size, the aligned_offset could in theory still add on an extra page to the buffer needed to read in the data aligned.
+                    // So offset % ps represents the starting point in the buffer where the requested data is (see the second-to-last line).
+                    // If that point plus length runs over ps, we need another ps worth of data.
+                    let buf_len = if length < ps { ((offset % ps) + length).next_multiple_of(ps) } else { length.next_multiple_of(ps) + ps };
+                    // Because we can't `read_exact` more bytes than are in the file, we need to use these libc calls to do O_DIRECT reading.
+                    let mut buf_ptr: *mut std::ffi::c_void = std::ptr::null_mut();
+                    let ret = unsafe { libc::posix_memalign(&mut buf_ptr, ps,  buf_len) };
+                    assert!(ret == 0, "posix_memalign failed during O_DIRECT reading");
+                    let buf = unsafe { std::slice::from_raw_parts_mut(buf_ptr.cast::<u8>(), buf_len) };
+                    let read_bytes = unsafe { libc::pread(fd, buf_ptr, buf_len, i64::try_from(aligned_offset).unwrap()) };
+                    assert!(read_bytes >= 0, "pread failed during O_DIRECT reading");
+
+                    let start_in_buf = offset - aligned_offset;
+                    let last_bytes = buf[start_in_buf..(start_in_buf + length)].to_vec();
+                    // Free the unused memory
+                    unsafe { libc::free(buf_ptr) };
+                    return Ok(Bytes::from(last_bytes));
```

Comment:
The output of this method is [`bytes::Bytes`](https://docs.rs/bytes/latest/bytes/struct.Bytes.html), which can be a subset of allocated memory. I think it should be feasible to avoid a copy entirely, but I'm not 100% sure. For example, see the `bytes_aligned` function at the top of this file.